src/version.ts

```diff
@@ -0,0 +1,318 @@
+ * @param project: The Julia project directory or file to use when determining
+ * Julia compatibility with a project.
+ * @param options: The `ifMissing` option controls the behavior of this
+ * function when a version specifier cannot be resolved.
+ * @returns A list of resolved versions
+ * @throws Error if a version specifier doesn't resolve to any available
+ * Julia release
+ */
+export async function resolveVersions(
+  versionSpecifiers: Array<string>,
+  project: string = ".",
+  options?: { ifMissing: string }
+): Promise<Array<string | null>> {
+  // Determine the Julia compat ranges as specified by the Project.toml only
+  // for aliases that require them.
+  let juliaCompatRange: string = ""
+  if (versionSpecifiers.includes("min")) {
+    const juliaProjectFile = getJuliaProjectFile(project)
+    const juliaProjectToml = toml.parse(
+      fs.readFileSync(juliaProjectFile).toString()
+    )
+
+    // Extract the compat range for the "julia" entry and convert it to the
+    // node semver range syntax.
+    juliaCompatRange = getJuliaCompatRange(juliaProjectToml)
+    core.debug(`Julia project compatibility range: ${juliaCompatRange}`)
+  }
+
+  const availableVersions = Object.keys(await fetchJuliaVersionsJson())
+
+  const resolvedVersions = new Array<string | null>()
+  for (const versionSpecifier of versionSpecifiers) {
+    let resolvedVersion: string | null
+
+    // Nightlies are not included in the versions.json file
+    const nightlyMatch = /^(?:(\d+\.\d+)-)?nightly$/.exec(versionSpecifier)
+    if (nightlyMatch && nightlyMatch[1]) {
+      const url = getNightlyUrl(DEFAULT_NIGHTLY_PLATFORM, nightlyMatch[1])
+      resolvedVersion = (await urlExists(url)) ? versionSpecifier : null
+    } else if (nightlyMatch) {
+      // Skip URL check for "nightly" as it should always be available
+      resolvedVersion = versionSpecifier
+    } else {
+      resolvedVersion = resolveVersion(
+        versionSpecifier,
+        availableVersions,
+        juliaCompatRange
+      )
+    }
+
+    core.debug(`${versionSpecifier} -> ${resolvedVersion}`)
+
+    if (resolvedVersion) {
+      resolvedVersions.push(resolvedVersion)
+    } else if (options?.ifMissing === "warn") {
+      core.warning(
+        `No Julia version exists matching specifier: "${versionSpecifier}"`
+      )
+    } else {
+      throw new Error(
+        `No Julia version exists matching specifier: "${versionSpecifier}"`
+      )
+    }
+  }
+
+  return resolvedVersions
+}
+
+/**
+ * Fetch and parse the Julia versions.json file.
+ *
+ * @returns The content of the downloaded versions.json file as object.
+ */
+export async function fetchJuliaVersionsJson(): Promise<JuliaVersionsJson> {
+  // Occasionally the connection is reset for unknown reasons
+  // In those cases, retry the download
+  const versionsFile = await retry(
+    async () => {
+      const response = await fetch(VERSIONS_JSON_URL)
+      return response.text()
+    },
+    {
+      retries: 5,
```

Comment:
Might be worth considering a back-off here

Reply:
The `retry` function uses an exponential backoff by default: https://www.npmjs.com/package/async-retry